### PR TITLE
Adds new helper: use_tool, shakes things up in tool code

### LINKED ISSUE #35198: mech construction :b:roke
Issue reported from  Round ID: 82893 (/tg/Station Bagil [ENGLISH] [US-EAST] [100% FREE LAG])

Reporting client version: 511

[Round ID]: # (If you discovered this issue from playing tgstation hosted servers:)
[Round ID]: # (**INCLUDE THE ROUND ID**)
[Round ID]: # (It can be found in the Status panel or retrieved from https://atlantaned.space/statbus/round.php ! The round id let's us look up valuable information and logs for the round the bug happened.)

[Testmerges]: # (If you believe the issue to be caused by a test merge [OOC tab -> Show Server Revision], report it in the pull request's comment section instead.)

[Reproduction]: # (Explain your issue in detail, including the steps to reproduce it. Issues without proper reproduction steps or explanation are open to being ignored/closed by maintainers.)

[For Admins]: # (Oddities induced by var-edits and other admin tools are not necessarily bugs. Verify that your issues occur under regular circumstances before reporting them.)

you cant use tools on mechs to build them caused by testmerged pr

### PR DESCRIPTION
**Pull request by ACCount12**
This PR replaces most of the `do_after` tool delays and the surrounding checks with just two procs: `tool_start_check` and `use_tool`.

**/obj/item/proc/tool_start_check(user, amount=0)**
Performs checks to determine if a tool is suitable to start an operation. Checks if things like welding tools or stacks have `amount` fuel/sheets in them, also handles tool-specific checks like welding tools being on and displays error messages to the user if any such checks fail. Applies tool-specifc effects such as welding tools blinding people.

It's safe to skip this proc for "dumb" tools (no possible checks) such as screwdrivers.

**/obj/item/proc/use_tool(target, user, delay, amount=0, volume=0, extra_checks)**
Actual tool usage. Handles the `do_after` delay (reduced by tool's `toolspeed`), runs `extra_checks`, amount checks and tool-specific checks on every tick of `do_after`, plays tool sound at `volume` at the beginning and at the end of tool use, uses up tool's fuel/sheets. Also runs `tool_start_check` if input delay is 0, as there is no reason to separate `use_tool` and `tool_start_check` unless you want to display a message between the two.

Example:
```
/obj/item/example/welder_act(mob/living/user, obj/item/I)
	if(!I.tool_start_check(user, amount=1))
		return TRUE

	user.visible_message("[user] is welding [src].",
		"<span class='notice'>You begin welding [src]...</span>",
		"<span class='italics'>You hear welding.</span>")

	if(I.use_tool(src, user, 40, volume=50, amount=1))
		to_chat(user, "<span class='notice'>You weld [src].</span>")
		<actual effect>
	return TRUE
```
Both procs are safe to use on any item, including all the regular tools. Amount checks work on all sheets, welding tools and plasma cutters (plasma cutters being common welder replacements).

This PR also throws in some code improvements I made when I was applying `use_tool`:
* Fixes some logic errors in construction/deconstruction procs.
* Multiple deconstruction procs now use `drop_location()`.
* Multiple deconstruction procs no longer manually set the amount on dropped stacks.
* Some simple `attack_by` procs that only used one tool are ported to `tool_act`.
* Some snowflake fastening/unfastening operations are changed to `default_unfasten_wrench`.

Test merge results:

Closes #35198.

### CODE REVIEW

File: code/game/machinery/computer/_computer.dm [hunk 1]
```diff
@@ -69,14 +69,13 @@
...
 
-/obj/machinery/computer/attackby(obj/item/I, mob/user, params)
-	if(istype(I, /obj/item/screwdriver) && circuit && !(flags_1&NODECONSTRUCT_1))
-		playsound(src.loc, I.usesound, 50, 1)
-		to_chat(user, "<span class='notice'> You start to disconnect the monitor...</span>")
-		if(do_after(user, 20*I.toolspeed, target = src))
+/obj/machinery/computer/screwdriver_act(mob/living/user, obj/item/I)
+	if(circuit && !(flags_1&NODECONSTRUCT_1))
+		to_chat(user, "<span class='notice'>You start to disconnect the monitor...</span>")
+		if(I.use_tool(src, user, 40, volume=50))
```
Comment (ShizCalev) [inline code/game/machinery/computer/_computer.dm]: Changing behavior here, no? Delay was 20

File: code/game/objects/structures/mirror.dm [hunk 2]
```diff
@@ -61,19 +61,17 @@
...
-					broken = 0
-					icon_state = initial(icon_state)
-					desc = initial(desc)
+		if(!broken)
+			return
+
+		if(!I.tool_start_check(user, amount=0))
+			return
+		to_chat(user, "<span class='notice'>You begin repairing [src]...</span>")
+		if(I.use_tool(src, user, 20, volume=50))
```
Comment (ShizCalev) [inline code/game/objects/structures/mirror.dm]: 10 -> 20?

File: code/game/objects/structures/watercloset.dm [hunk 3]
```diff
@@ -244,7 +244,7 @@
 		to_chat(user, "<span class='notice'>The water temperature seems to be [watertemp].</span>")
 	if(istype(I, /obj/item/wrench))
 		to_chat(user, "<span class='notice'>You begin to adjust the temperature valve with \the [I]...</span>")
-		if(do_after(user, 50*I.toolspeed, target = src))
+		if(I.use_tool(src, user, 40))
```
Comment (ShizCalev) [inline code/game/objects/structures/watercloset.dm]: 50 -> 40?

File: code/game/objects/structures/watercloset.dm [hunk 4]
```diff
@@ -610,15 +610,15 @@
 		if(anchored)
 			playsound(src.loc, W.usesound, 100, 1)
 			user.visible_message("<span class='warning'>[user] unscrews [src] from the floor.</span>", "<span class='notice'>You start to unscrew [src] from the floor...</span>", "You hear rustling noises.")
-			if(do_after(user, 50*W.toolspeed, target = src))
+			if(W.use_tool(src, user, 40))
```
Comment (ShizCalev) [inline code/game/objects/structures/watercloset.dm]: 50 -> 40

File: code/game/objects/structures/watercloset.dm [hunk 5]
```diff
@@ -610,15 +610,15 @@
...
 				if(!anchored)
 					return
 				anchored = FALSE
 				to_chat(user, "<span class='notice'>You unscrew [src] from the floor.</span>")
 		else
 			playsound(src.loc, W.usesound, 100, 1)
 			user.visible_message("<span class='warning'>[user] screws [src] to the floor.</span>", "<span class='notice'>You start to screw [src] to the floor...</span>", "You hear rustling noises.")
-			if(do_after(user, 50*W.toolspeed, target = src))
+			if(W.use_tool(src, user, 40))
```
Comment (ShizCalev) [inline code/game/objects/structures/watercloset.dm]: 50 -> 40

File: code/game/objects/structures/watercloset.dm [hunk 6]
```diff
@@ -627,7 +627,7 @@
 		if(!anchored)
 			playsound(src.loc, W.usesound, 100, 1)
 			user.visible_message("<span class='warning'>[user] cuts apart [src].</span>", "<span class='notice'>You start to cut apart [src].</span>", "You hear cutting.")
-			if(do_after(user, 50*W.toolspeed, target = src))
+			if(W.use_tool(src, user, 40))
```
Comment (ShizCalev) [inline code/game/objects/structures/watercloset.dm]: 50 -> 40

**Review by ShizCalev — CHANGES_REQUESTED** [review]

File: code/game/machinery/camera/camera_assembly.dm [hunk 7]
```diff
@@ -128,15 +128,14 @@
...
 	to_chat(user, "<span class='notice'>You start to weld \the [src]...</span>")
-	playsound(src.loc, WT.usesound, 50, 1)
-	if(do_after(user, 20*WT.toolspeed, target = src))
-		if(WT.isOn())
-			playsound(loc, 'sound/items/welder2.ogg', 50, 1)
-			return 1
+	playsound(src.loc, W.usesound, 50, 1)
+	if(W.use_tool(src, user, 20))
+		playsound(loc, 'sound/items/welder2.ogg', 50, 1)
+		return 1
```
Comment (ShizCalev) [inline code/game/machinery/camera/camera_assembly.dm]: TRUE

File: code/game/turfs/simulated/floor/misc_floor.dm [hunk 8]
```diff
@@ -196,19 +196,14 @@
...
-			return 0
+		return TRUE
+	user.visible_message("<span class='notice'>[user] begins slowly prying up [src]...</span>", "<span class='notice'>You begin painstakingly prying up [src]...</span>")
+	if(I.use_tool(src, user, 70, volume=80))
 		user.visible_message("<span class='notice'>[user] pries up [src]!</span>", "<span class='notice'>You pry up [src]!</span>")
-		playsound(src, I.usesound, 80, 1)
 		make_plating()
-		return 1
-	return ..()
+	return TRUE
```
Comment (ShizCalev) [inline code/game/turfs/simulated/floor/misc_floor.dm]: Shouldn't this be under the if?

File: code/game/turfs/simulated/wall/reinf_walls.dm [hunk 9]
```diff
@@ -132,34 +118,31 @@
...
-							return 1
-						d_state = COVER
-						update_icon()
-						to_chat(user, "<span class='notice'>The metal cover has been welded securely to the frame.</span>")
+				if(!W.tool_start_check(user, amount=0))
+					return
+				to_chat(user, "<span class='notice'>You begin welding the metal cover back to the frame...</span>")
+				if(W.use_tool(src, user, 60, volume=100))
+					if(!istype(src, /turf/closed/wall/r_wall) || d_state != CUT_COVER)
+						return 1
```
Comment (ShizCalev) [inline code/game/turfs/simulated/wall/reinf_walls.dm]: TRUE

File: code/game/turfs/simulated/wall/reinf_walls.dm [hunk 10]
```diff
@@ -216,26 +186,24 @@
...
-						d_state = SUPPORT_RODS
-						update_icon()
-						to_chat(user, "<span class='notice'>You weld the support rods back together.</span>")
-					return 1
+				if(!W.tool_start_check(user, amount=0))
+					return
+				to_chat(user, "<span class='notice'>You begin welding the support rods back together...</span>")
+				if(W.use_tool(src, user, 100, volume=100))
+					if(!istype(src, /turf/closed/wall/r_wall) || d_state != SHEATH)
+						return 1
```
Comment (ShizCalev) [inline code/game/turfs/simulated/wall/reinf_walls.dm]: TRUE

File: code/game/turfs/simulated/wall/reinf_walls.dm [hunk 11]
```diff
@@ -216,26 +186,24 @@
...
+					d_state = SUPPORT_RODS
+					update_icon()
+					to_chat(user, "<span class='notice'>You weld the support rods back together.</span>")
+				return 1
```
Comment (ShizCalev) [inline code/game/turfs/simulated/wall/reinf_walls.dm]: Indent?

File: code/game/turfs/simulated/walls.dm [hunk 12]
```diff
@@ -219,27 +222,17 @@
...
-			if(!iswallturf(src) || !user || QDELETED(W) || QDELETED(T))
-				return TRUE
-			if((user.loc == T) && (user.get_active_held_item() == W))
+		if(W.use_tool(src, user, slicing_duration, volume=100))
+			if(iswallturf(src))
 				to_chat(user, "<span class='notice'>You remove the outer plating.</span>")
 				dismantle_wall()
-				visible_message("The wall was sliced apart by [user]!", "<span class='italics'>You hear metal being sliced apart.</span>")
-				return TRUE
+			return TRUE
```
Comment (ShizCalev) [inline code/game/turfs/simulated/walls.dm]: indent, no?

File: code/game/machinery/computer/_computer.dm [hunk 13]
```diff
@@ -69,14 +69,13 @@
...
 			deconstruct(TRUE, user)
-	else
-		return ..()
+	return TRUE
```
Comment (ShizCalev) [inline code/game/machinery/computer/_computer.dm]: indent?

**Comment by ACCount12** [comment]
I remember trying to standardize action times to make them follow `20 - 40 - 60 - too fucking long` pattern. Guess that's how that went.

**Review by ShizCalev — DISMISSED** [review]

File: code/game/machinery/computer/_computer.dm (hunk 13 above)
Comment (ACCount12) [reply code/game/machinery/computer/_computer.dm]: Nope.

File: code/game/turfs/simulated/floor/misc_floor.dm (hunk 8 above)
Comment (ACCount12) [reply code/game/turfs/simulated/floor/misc_floor.dm]: No.

File: code/game/turfs/simulated/wall/reinf_walls.dm (hunk 11 above)
Comment (ACCount12) [reply code/game/turfs/simulated/wall/reinf_walls.dm]: No.

File: code/game/turfs/simulated/walls.dm (hunk 12 above)
Comment (ACCount12) [reply code/game/turfs/simulated/walls.dm]: No, and no to all of the above. Return there just stops attack_by from going through, and you don't want it to go through in most tool interactions.

**Comment by ACCount12** [comment]
Can I please do not replace every single `1` I touched with `TRUE`? I try to leave the code in a better state, but this is getting ridiculous.

**Comment by arsserpentarium** [comment]
Or from atom/moveable.

**Comment by uraniummeltdown** [comment]
@ACCount12 It is annoying dealing with 0/1 TRUE/FALSE but please don't give up on PRs because of that

File: code/game/machinery/camera/camera_assembly.dm [hunk 14]
```diff
@@ -128,15 +128,14 @@
 	qdel(src)
 	return TRUE
 
-/obj/structure/camera_assembly/proc/weld(obj/item/weldingtool/WT, mob/living/user)
-	if(!WT.remove_fuel(0, user))
+/obj/structure/camera_assembly/proc/weld(obj/item/weldingtool/W, mob/living/user)
+	if(!W.tool_start_check(user, amount=0))
 		return 0
```
Comment (nicbn) [inline code/game/machinery/camera/camera_assembly.dm]: FALSE

**Comment by ShizCalev** [comment]
Is changing 3 lines in a PR that changes 1535 lines really too much?

**Comment by ACCount12** [comment]
Not really, but as soon as I fix those 3 lines, you are going to have an another review pass and find at least 3 more. Repeat until my head explodes from sheer frustration.

**Comment by ACCount12** [comment]
Can I get a test merge now?

**Comment by Cyberboss** [comment]
Does this handle starting a weld then flipping the helmet up?

**Comment by ACCount12** [comment]
Nope. The old system didn't, this doesn't either. The purpose of this is to make adding new tools or writing new construction code easier, not to screw players over.

**Comment by arsserpentarium** [comment]
Can I ask you to test it userless?

**Comment by ACCount12** [comment]
It was test merged, one bug was found and fixed in the process. Ready to merge I think.

**Pull request merged**